File: julia/src/benchmarks/Binarytrees.jl
mutable struct BinarytreesObj <: AbstractBenchmark
    n::Int64
    result::UInt32
end

function BinarytreesObj()
    n = Helper.config_i64("BinarytreesObj", "depth")
    return BinarytreesObj(n, UInt32(0))
end

name(b::BinarytreesObj)::String = "BinarytreesObj"

mutable struct TreeNodeObj
    left::Union{Nothing,TreeNodeObj}
    right::Union{Nothing,TreeNodeObj}
    item::Int64

    function TreeNodeObj(item::Int64, depth::Int64)
        node = new(nothing, nothing, item)
        if depth > 0
            shift = 1 << (depth - 1)
            node.left = TreeNodeObj(item - shift, depth - 1)
            node.right = TreeNodeObj(item + shift, depth - 1)
        end
        return node
    end
end

function sum(node::TreeNodeObj)::Int64
    total = node.item + 1

    if node.left !== nothing
        total += sum(node.left)
    end
    if node.right !== nothing
        total += sum(node.right)
    end
    return total
end

function run(b::BinarytreesObj, iteration_id::Int64)
    tree = TreeNodeObj(0, b.n)
    b.result += Helper.to_u32(sum(tree))

end

checksum(b::BinarytreesObj)::UInt32 = b.result

mutable struct TreeNodeArena
    item::Int64
    left::Int64
    right::Int64

    function TreeNodeArena(item::Int64)
        return new(item, -1, -1)
    end
end

mutable struct BinarytreesArena <: AbstractBenchmark
    n::Int64
    result::UInt32

end

function BinarytreesArena()
    n = Helper.config_i64("BinarytreesArena", "depth")
    return BinarytreesArena(n, UInt32(0))
end

name(b::BinarytreesArena)::String = "BinarytreesArena"

function build!(nodes::Vector{TreeNodeArena}, item::Int64, depth::Int64)::Int64
    idx = length(nodes) + 1
    push!(nodes, TreeNodeArena(item))

    if depth > 0
        shift = 1 << (depth - 1)
        left_idx = build!(nodes, item - shift, depth - 1)
        right_idx = build!(nodes, item + shift, depth - 1)
        nodes[idx].left = left_idx
        nodes[idx].right = right_idx
    end

    return idx
end

function sum(nodes::Vector{TreeNodeArena}, idx::Int64)::Int64
    node = nodes[idx]
    total = node.item + 1

    if node.left >= 1
        total += sum(nodes, node.left)
    end
    if node.right >= 1
        total += sum(nodes, node.right)
    end

    return total
end

function run(b::BinarytreesArena, iteration_id::Int64)

    nodes = TreeNodeArena[]
    root_idx = build!(nodes, 0, b.n)
    b.result += Helper.to_u32(sum(nodes, root_idx))

end

checksum(b::BinarytreesArena)::UInt32 = b.result

File: nim/src/benchmarks/binarytrees.nim
import std/math
import ../benchmark

type
  TreeNodeObj = ref object
    left, right: TreeNodeObj
    item: int

proc newTreeNodeObj(item: int, depth: int): TreeNodeObj =
  result = TreeNodeObj(item: item)
  if depth > 0:
    let shift = 1 shl (depth - 1)
    result.left = newTreeNodeObj(item - shift, depth - 1)
    result.right = newTreeNodeObj(item + shift, depth - 1)

proc sum(self: TreeNodeObj): uint32 =
  var total = uint32(self.item) + 1'u32
  if not self.left.isNil:
    total += self.left.sum()
  if not self.right.isNil:
    total += self.right.sum()
  return total

type
  BinarytreesObj* = ref object of Benchmark
    n: int64
    resultVal: uint32

proc newBinarytreesObj(): Benchmark =
  BinarytreesObj()

method name(self: BinarytreesObj): string = "Binarytrees::Obj"

method prepare(self: BinarytreesObj) =
  self.n = self.config_val("depth")
  self.resultVal = 0

method run(self: BinarytreesObj, iteration_id: int) =
  let tree = newTreeNodeObj(0, self.n.int)
  self.resultVal += tree.sum()

method checksum(self: BinarytreesObj): uint32 =
  self.resultVal

registerBenchmark("Binarytrees::Obj", newBinarytreesObj)

type
  TreeNodeArena = object
    item: int
    left: int
    right: int

  TreeArena = object
    nodes: seq[TreeNodeArena]

proc initTreeArena(): TreeArena =
  result.nodes = newSeq[TreeNodeArena]()

proc build(self: var TreeArena, item: int, depth: int): int =
  let idx = self.nodes.len
  self.nodes.add(TreeNodeArena(item: item, left: -1, right: -1))

  if depth > 0:
    let shift = 1 shl (depth - 1)
    let leftIdx = self.build(item - shift, depth - 1)
    let rightIdx = self.build(item + shift, depth - 1)
    self.nodes[idx].left = leftIdx
    self.nodes[idx].right = rightIdx

  return idx

proc sum(self: TreeArena, idx: int): uint32 =
  let node = self.nodes[idx]
  var total = uint32(node.item) + 1'u32

  if node.left >= 0:
    total += self.sum(node.left)
  if node.right >= 0:
    total += self.sum(node.right)

  return total

type
  BinarytreesArena* = ref object of Benchmark
    n: int64
    resultVal: uint32

proc newBinarytreesArena(): Benchmark =
  result = BinarytreesArena()

method name(self: BinarytreesArena): string = "Binarytrees::Arena"

method prepare(self: BinarytreesArena) =
  self.n = self.config_val("depth")
  self.resultVal = 0

method run(self: BinarytreesArena, iteration_id: int) =
  var arena = initTreeArena()
  let rootIdx = arena.build(0, self.n.int)
  self.resultVal += arena.sum(rootIdx)

method checksum(self: BinarytreesArena): uint32 =
  self.resultVal

registerBenchmark("Binarytrees::Arena", newBinarytreesArena)
